dom.go

```diff
@@ -252,19 +268,59 @@ func (h *HTML) reconcile(prev *HTML) []Mounter {
 		h.node.Set("innerHTML", h.innerHTML)
 	}
 
-	return h.reconcileChildren(prev, nil)
+	return h.reconcileChildren(prev)
 }
 
 // reconcileChildren reconciles children of the current HTML against a previous
 // render's DOM nodes.
-func (h *HTML) reconcileChildren(prev *HTML, insertBefore *HTML) (pendingMounts []Mounter) {
-	// TODO better list element reuse
+func (h *HTML) reconcileChildren(prev *HTML) (pendingMounts []Mounter) {
+	var keyed bool
 	for i, nextChild := range h.children {
-		// TODO(pdf): Add tests for node equality
-		new := h.node != prev.node
-		if i >= len(prev.children) || new {
-			if nextChildList, ok := nextChild.(List); ok {
-				nextChildList.reconcile(h.node, insertBefore, nil)
+		// Massage concrete type if necessary.
+		switch v := nextChild.(type) {
+		case *HTML:
+			if v == nil {
+				nextChild = nil
+				h.children[i] = nextChild
+			}
+		case List:
+			nextChild = childList{html: &HTML{children: v}}
```

Comment:
Document. There is a reason why we are replacing `List` types with our own `childList` type, but I do not know what it is.